pkg/stream/structs.go

```diff
@@ -0,0 +1,100 @@
+package stream
+
+import (
+	"github.com/blocknative/dreamboat/pkg/structs"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/flashbots/go-boost-utils/types"
+)
+
+func ToStreamBlock(block *structs.BlockAndTrace, isCache bool, source string) *StreamBlock {
+	var transactions [][]byte
+	for _, tx := range block.Payload.Data.Transactions {
+		transactions = append(transactions, tx)
+	}
+
+	return &StreamBlock{
+		Source:  source,
+		IsCache: isCache,
+		Block: &PayloadAndTrace{
+			Trace: &SignedBidTrace{
+				Signature: block.Trace.Signature[:],
+				Message: &BidTrace{
+					Slot:                 block.Trace.Message.Slot,
+					ParentHash:           block.Trace.Message.ParentHash[:],
+					BlockHash:            block.Trace.Message.BlockHash[:],
+					BuilderPubkey:        block.Trace.Message.BuilderPubkey[:],
+					ProposerPubkey:       block.Trace.Message.ProposerPubkey[:],
+					ProposerFeeRecipient: block.Trace.Message.ProposerFeeRecipient[:],
+					GasLimit:             block.Trace.Message.GasLimit,
+					GasUsed:              block.Trace.Message.GasUsed,
+					Value:                block.Trace.Message.Value[:],
+				},
+			},
+			Payload: &PayloadWithVersion{
+				Version: string(block.Payload.Version),
+				Payload: &ExecutionPayload{
+					ParentHash:    block.Payload.Data.ParentHash[:],
+					FeeRecipient:  block.Payload.Data.FeeRecipient[:],
+					StateRoot:     block.Payload.Data.StateRoot[:],
+					ReceiptsRoot:  block.Payload.Data.ReceiptsRoot[:],
+					LogsBloom:     block.Payload.Data.LogsBloom[:],
+					Random:        block.Payload.Data.Random[:],
+					BlockNumber:   block.Payload.Data.BlockNumber,
+					GasLimit:      block.Payload.Data.GasLimit,
+					GasUsed:       block.Payload.Data.GasUsed,
+					Timestamp:     block.Payload.Data.Timestamp,
+					ExtraData:     block.Payload.Data.ExtraData,
+					BaseFeePerGas: block.Payload.Data.BaseFeePerGas[:],
+					BlockHash:     block.Payload.Data.BlockHash[:],
+					Transactions:  transactions,
+				},
+			},
+		},
+	}
+}
+
+func ToBlockAndTrace(streamBlock *StreamBlock) *structs.BlockAndTrace {
+	block := structs.BlockAndTrace{
+		Trace: &types.SignedBidTrace{
+			Message: &types.BidTrace{},
+		},
+		Payload: &types.GetPayloadResponse{
+			Data: &types.ExecutionPayload{},
+		},
+	}
+
+	// copy trace
+	copy(block.Trace.Signature[:], streamBlock.Block.Trace.Signature)
```

Comment:
Is the copy required?  I'm generally in favor of defensive copies on slices, but there's a _lot_ of them, so I start to worry about performance implications.

This seems to be called on each stream element, with O(n) complexity.  Do we have an idea of stream volume?

Reply:
The copy is needed because unfortunately one of them is an array, while the other is a slice. Unless there is a way I don't know about

Reply:
Ok, that makes sense.  Unless we're expecting very large data streams, I'm okay to ship this.  Just keep an eye on performance, in case this introduces a sudden severe slowdown.  We might even consider adding a prometheus metric to track this, if it's not too much work.